### DISCUSSION TITLE
Why NodeJS store source code in memory?

### DISCUSSION THREAD

**Discussion by Luchanso** [nodejs/help#5049] [Q&A] [answered]
Hey guys, I took a heap snapshot of empty application:
```js
// index.js
setTimeout(function () {}, 100000000);
```
```sh
node --inspect index
```

And I don't understand why NodeJS store source code in strings? It is allocate ~35% memory (or 1 mb memory for empty application) with 29% compiled code
<img width="1046" alt="image" src="https://user-images.githubusercontent.com/2098777/223369935-1bafc611-5530-40b2-b955-c231c8fb3a23.png">
<img width="1048" alt="image" src="https://user-images.githubusercontent.com/2098777/223370503-31d70a6b-fcea-45dd-8ec7-82c42ec9de88.png">

Or is it a question for V8 & TurboFun?
This is problem for large application, when code base is large it can take from 25 mb to 300 mb of memory for these strings

**Comment by bnoordhuis** [comment, marked as answer]
It's pretty much required by the language. Calling `.toString()` on a function must return the source text.

SpiderMonkey (the JS engine in Firefox) at one time reconstructed the source text from bytecode but that had so many bugs and edge cases, it never really worked right.

They eventually gave up trying to make it work and switched to retaining source in memory, just like all other major JS engines.

**Comment by Luchanso** [comment]
@bnoordhuis thank you for answer, maybe there is a way to delete or disable this behavior? (for experiments)

**Reply by bnoordhuis** [reply]
Not in V8, it's baked in at a pretty fundamental level.
